### ISSUE TITLE
Convnet CUDA problem (OSX 10.10)

### ISSUE THREAD

**Issue by mbertini** [liuliu/ccv#122]
I'm trying to use the CUDA acceleration on a MacBookPro (NVIDIA GeForce GT 750M) with CUDA 6.5 on OS X 10.10.
As soon as I change:
    ccv_convnet_t\* convnet = ccv_convnet_read(0, argv[2]);
to
    ccv_convnet_t\* convnet = ccv_convnet_read(1, argv[2]);

in bin/cnnclassify.c

I get at run time the following assertion:
    Assertion failed: (layers[i].w), function _cwc_convnet_alloc_layers, file cuda/cwc_convnet.cu, line 158.

Using the CPU everything is OK. Compilation seems OK with -D HAVE_CUDA showing that it was correctly found.

**Comment by liuliu** [comment]
Thanks. Can you let me know how much graphics memory the MBP has?

You can print that out how much GPU memory ccv needs before the assertion: https://github.com/liuliu/ccv/blob/unstable/lib/cuda/cwc_convnet.cu#L157

**Comment by mbertini** [comment]
I've tried to get some GPU info with a small test program (changing one of the NVIDIA CUDA SDK 6.5 samples - i.e. samples/1_Utilities/bandwidthTest/bandwidthTest.cu) and I get the following:

size_t cudaFree, cudaTotal;
cudaError_t cuErr;
cuErr  = cudaMemGetInfo(&cudaFree, &cudaTotal);
printf("CudaFree: %lu - CUDA Total: %lu - CUDAError: %d\n", cudaFree, cudaTotal, cuErr);

CudaFree: 660914176 - CUDA Total: 2147024896 - CUDAError: 0

If I add the same code to cwd_convnet.cu and cnnclassify.c I get:

CudaFree: 0 - CUDA Total: 0 - CUDAError: 63  
The error code (63) seems to be: cudaErrorOperatingSystem, i.e. a OS call that has failed. 

I'm sorry but I've really never programmed anything in CUDA, so I don't really know where to instrument the code... If you have any suggestion I can try again.

**Comment by mbertini** [comment]
I've added the following function to cnnclassify.c:

void CUDAMem() {
    int currentDevice = 0;
    struct cudaDeviceProp deviceProp;
    cudaError_t error_id = cudaGetDeviceProperties(&deviceProp, currentDevice);
    printf(" Device %d: %s\n", currentDevice, deviceProp.name);
    size_t cudaFree, cudaTotal;
    cudaError_t cuErr;
    cuErr  = cudaMemGetInfo(&cudaFree, &cudaTotal);
    printf("1) CudaFree: %lu - CUDA Total: %lu - CUDAError: %d\n", cudaFree, cudaTotal, cuErr);
    fflush(stdout);
}

then I've changed the first lines of main() in the same file:

int main(int argc, char*\* argv)
{
    CUDAMem();
    return 100;
    assert(argc >= 3);
...
....
}

If I return immediately I get something like:

nirvana:bin bertini$ ./cnnclassify ../samples/dex.png ../samples/image-net-2012.sqlite3
 Device 0: GeForce GT 750M
1) CudaFree: 215724032 - CUDA Total: 2147024896 - CUDAError: 0
nirvana:bin bertini$ echo $?
100

If I remove the return I get:

nirvana:bin bertini$ ./cnnclassify ../samples/dex.png ../samples/image-net-2012.sqlite3
 Device 0: GeForce GT 750M
1) CudaFree: 0 - CUDA Total: 0 - CUDAError: 63
GPU(convnet)->stats.memory_usage: 56832
CudaFree: 0 - CUDA Total: 0 - CUDAError: 63
Assertion failed: (layers[i].w), function _cwc_convnet_alloc_layers, file cuda/cwc_convnet.cu, line 163.
Abort trap: 6

where the second CudaFree and CUDA Total are printed from cwc_convnet.cu, where I added a similar code.

**Comment by liuliu** [comment]
This is very interesting, I am going to get a MBP to reprod this. Thanks for reporting.

**Comment by mbertini** [comment]
I've just tried using the new CUDA SDK 7.0 RC but there's still the same behavior.

**Comment by liuliu** [comment]
Yes, it should be device capacity issues rather than CUDA versions. Can you be more specific which NVIDIA card on your MBP? I may need to hunt down that specific card.

**Comment by mbertini** [comment]
It's a NVIDIA GeForce GT 750M with 2048 MB

**Comment by kylemcdonald** [comment]
i can report the same issue with the exact same setup (10.10, 750M/2GB, CUDA 7.0). i also tried to run the `cifar-10` example because i thought it would be a smaller network, but it fails when creating the network with the same error.

**Comment by liuliu** [comment]
Yes, I am aware of this problem. But unable to hunt down a MBP with given graphic card to debug. Will move to CUDNN for most of the convnet computations and hope would resolve this problem when that is done.

**Comment by guilhermeba** [comment]
I'm trying to compile Convnet in OS X 10.10, what's the right directories in "build.sh", for instance, I can't find NVIDIA_GPU_Computing_SDK in my installation. Can you help me?
